### PR TITLE
8269556: sun/tools/jhsdb/JShellHeapDumpTest.java fails with RuntimeException 'JShellToolProvider' missing from stdout/stderr

### PR DESCRIPTION
**Pull request by plummercj**
The test searches for "JShellToolProvider" in the main thread's stack trace, which is pulled from an SA heap dump. Typically the main thread is blocked in Object.wait(), so SA can determine its stack trace. However, the wait has a 100ms timeout, so the thread does periodically wake up and does a few things before waiting again. If SA tries to get the stack trace while the thread is active, it may not be able to, and this causes the test to fail. I determined this only happens about 1 in 5000-10000 runs. So as a fix I'm allowing the test to do one retry. That should make it extremely unlikely that we ever see this failure again. I also bumped up the amount of time the test waits before doing the heap dump from 2 seconds to 4 just to make absolutely sure the main thread is fully started before doing the heap dump.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269556](https://bugs.openjdk.java.net/browse/JDK-8269556): sun/tools/jhsdb/JShellHeapDumpTest.java fails with RuntimeException 'JShellToolProvider' missing from stdout/stderr

### Reviewers
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to 65e14883060878e3cb092d8e2fa51fa2f34e47af
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Alex Menkov](https://openjdk.java.net/census#amenkov) (@alexmenkov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6795/head:pull/6795` \
`$ git checkout pull/6795`

Update a local copy of the PR: \
`$ git checkout pull/6795` \
`$ git pull https://git.openjdk.java.net/jdk pull/6795/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6795`

View PR using the GUI difftool: \
`$ git pr show -t 6795`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6795.diff">https://git.openjdk.java.net/jdk/pull/6795.diff</a>

</details>

### CODE REVIEW

**Review by kevinjwalls — APPROVED** [review]
Looks good.

**Review by sspitsyn — APPROVED** [review]
Hi Chris,
It looks good in general. I feel the parameter 'retry' is a little bit confusing. It is possible to rename it to some to 'retriedOnce'. Another approach would be to name it 'allowRetry' and revert the meaning of boolean value. 
I leave it up to you to keep the original approach or make any change.
Thanks,
Serguei

**Comment by plummercj** [comment]
> Hi Chris,
> It looks good in general. I feel the parameter 'retry' is a little bit confusing. It is possible to rename it to some to 'retriedOnce'. Another approach would be to name it 'allowRetry' and revert the meaning of boolean value.
> I leave it up to you to keep the original approach or make any change.
> Thanks,
> Serguei

Ok. I changed it to allowRetry.

File: test/jdk/sun/tools/jhsdb/JShellHeapDumpTest.java [hunk 1]
```diff
@@ -97,21 +97,33 @@ public static void launch(String expectedMessage, String... toolArgs)
...
             // We only require JShellToolProvider to be in the output if we did the
             // short sleep. If we did not, the java process may not have executed far
             // enough along to even start the main thread.
             if (doSleep && !output.contains("JShellToolProvider")) {
-                throw new RuntimeException("'JShellToolProvider' missing from stdout/stderr");
+                // This check will very rarely fail due to not be able to get the stack trace
+                // of the main thread do to it actively executing. See JDK-8269556. We retry once
+                // if that happens. This failure is so rare that this should be enough to make it
+                // extremely unlikely that we ever see this test fail again for this reason.
+                if (allowRetry) {
```
Comment (alexmenkov) [inline test/jdk/sun/tools/jhsdb/JShellHeapDumpTest.java:112]: Looks like this should be !allowRetry

**Review by alexmenkov — CHANGES_REQUESTED** [review]

File: test/jdk/sun/tools/jhsdb/JShellHeapDumpTest.java (hunk 1 above)
Comment (plummercj) [reply test/jdk/sun/tools/jhsdb/JShellHeapDumpTest.java:112]: Yeah, I was just looking at the diff an noticed that, but you beat me to it. Will push another commit to fix.

**Review by alexmenkov — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**